### ISSUE TITLE
Update in PostgreSql, one bug and one issue

### ISSUE THREAD

**Issue by shishmak** [linq2db/linq2db#1549]
Hi!
could you please help me in my question?
I faced with the problem trying to write the “Update” function – I have 1 issue and I caught 1 bug. 
Can I ask you to help me to figure out how it works?

I need update table data by data from the source.

The PostgreSQL command syntax is below:
```sql
update [table] 
set [field1] = subq.[f1]
set [field2] = subq.[f2]
from  
 [sorce] as subq
Where subq.id = [table].id
(sql code 1)
```
or
```sql
update [table]
set [field1] = (select [f1] from [source] where id = [table] .id limit 1)
set [field2] = (select [f2] from [source] where id = [table] .id limit 1)
(sql code 2)
```

1.	First about I think is a bug.
C# code:
```cs
var items = db.Target
.Set(p => p.field1,
u => Sorce.Where(w => w.Id == u.Id).Select(s => s.f2).FirstOrDefault()
           );
    items
    .Update();
```
Linq2db converts this code to “sql code 2”.

If the Source is presented by table (db.Table), the query is working
But if the source is presented by “select with join” (db.Table1.Join(db.Table2, t1=>t1.key, t2 => t2.key, (t1,t2)=>new{t1,t2}) ), the query doesn't build correctly and request fails. For example, see the attachment “bug.cs”.
Could you please clarify this for me?

2.
There are two ways to write the update from the source command in the PostgreSQL.
I noted that the way 1 (“sql code 1” in my example) works much faster than the way 2 (“sql code 2” in my example). Moreover, the way 1 doesn’t depend on fields count needed to be updated.
For my task I need to create request by the way 1 for DBs that support this request (PostgreSQL, SQL Server, Sybase, etc.).
That’s my issue.
Could you please suggest me the best solution?

Attachment files: 
- [model.cs.txt](https://github.com/linq2db/linq2db/files/2778039/model.cs.txt)
 sample of model for my case.
 - [BugReport.cs.txt](https://github.com/linq2db/linq2db/files/2778041/BugReport.cs.txt)
 cs code leading to incorrect query, sql invalid query(form log), sql valid query (wrapped in regions)
 - [issue.sql.txt](https://github.com/linq2db/linq2db/files/2778042/issue.sql.txt)
sql code with execution plan and times for queries in "sql code 1" and "sql code 2" syntax. 

Thank you in advance!

### Environment details
linq2db version: *2.6.2.0*
Database Server: *E.g. PostgreSQL  9.6.9*
Database Provider: *E.g. Npgsql 3.2.5.0*
Operating system: *E.g. Windows 10*
Framework version: *.NET Framework 4.5.0*

**Comment by sdanyliv** [comment]
@shishmak, unfortunately such update is not working yet. I have prepared PR for this and we plan to publish it soon. https://github.com/linq2db/linq2db/pull/1278

**Comment by sdanyliv** [comment]
Please post your model for reproducing. I’ll test against this PR. But at first sign you have used wrong linq query to make such update.

**Comment by shishmak** [comment]
> Please post your model for reproducing. 
There are two attachment files in first post:
model.cs.txt - c# code of model for this case.
BugReport.cs.txt - C# code to reproducing my qwery.

>I’ll test against this PR. But at first sign you have used wrong linq query to make such update.
I will be glad, if you tell me a more correct way for this situation

**Comment by sdanyliv** [comment]
Will do, i have spent a lot of hours to make such updates working.

**Comment by sdanyliv** [comment]
@shishmak, I have rewritten your query and it works in PR's branch:
```cs
var query = 
    from tr in db.GetTable<billing_TempReading>()
    from drt in db.GetTable<billing_DevReadingType>().InnerJoin(drt =>
	    drt.Name == tr.ReadingTypeName && drt.DevTypeId == tr.Devtypeid)
    select new
    {
	    tr,
	    drt
    };

query
    .Set(p => p.tr.DevReadingTypeId, p => p.drt.Id)
    .Set(p => p.tr.Responsibility, p => p.drt.Responsibility)
    .Update();
```
generated SQL is:
```sql
UPDATE
	"TempReading"
SET
	"DevReadingTypeId" = drt."Id",
	"Responsibility" = drt."Responsibility"
FROM
	"DevReadingType" drt
WHERE
	drt."Name" = "TempReading"."ReadingTypeName" AND drt."DevTypeId" = "TempReading"."Devtypeid"
```

**Comment by shishmak** [comment]
@sdanyliv, it's good.

I am grateful for your reaction and the work done
But you tested "Update" where sorce only one table. it worked before. Interesting second part of the file bugReport (from line 63). When sorce is Join from 2 or more tables. My 3 comment in this issue about it.

Can you test this case? 
c# (bugReport.cs.txt line 63) 
```
var devs = db.Devices.Join(db.Devtypes, d => d.Devtypeid, dt => dt.Devtypeid, (d, dt) => new { d, dt });

db.TempReadings
	.Set(t => t.Devid, u => devs.Where(w => w.d.Sernum == u.DevSerNum && w.dt.GlobalType == u.DevGlobalType).Select(s => s.d.Devid).FirstOrDefault())
	.Set(t => t.Devtypeid, u => devs.Where(w => w.d.Sernum == u.DevSerNum && w.dt.GlobalType == u.DevGlobalType).Select(s => s.dt.Devtypeid).FirstOrDefault())
	.Update();
```
Thank you in advance.

**Comment by sdanyliv** [comment]
I confirm, query is not working. Will fix that also. But notice that my variant is more native for Postgres.

**Comment by MaceWindu** [comment]
Fixed by #1278
